### PR TITLE
[ENG-1662] feat: Add provider initialization

### PR DESCRIPTION
**Pull request by RajatPawar**
- Adds a function called `connector.NewConnector` to initialize a connector that can be later asserted as a read, write or metadata connector
- This function currently reuses the per-package params to get things to work. A later PR will include an example on how a connector can accept the `Parameters` directly.
- Adds a common parameter set called `connector.Parameters` which contains inputs that a connector generally demands for working.

Instead of adding validation logic to individual connectors, we can assume that the parameters are present, and validate them at entrypoints like `connector.NewConnector`. This will help a lot by enabling reusable common types.

### CODE REVIEW

**Comment by RajatPawar** [comment]
* **#1205** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1205?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1205?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by RajatPawar** [comment]
Added `nolint` directives to both the files for now to propose the PR

File: connector/new.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+	return generic.NewConnector(
+		provider,
+		generic.WithAuthenticatedClient(parameters.AuthenticatedClient),
+		generic.WithWorkspace(parameters.Workspace),
+	)
+}
+
+func newSalesforceConnector(parameters Parameters) (connectors.Connector, error) {
+	return salesforce.NewConnector(
+		salesforce.WithAuthenticatedClient(parameters.AuthenticatedClient),
```
Comment (RajatPawar) [inline connector/new.go:133]: Keeping these options as they are currently, will add a new PR to show how we can reduce the need to have per-package options

File: connector/params.go [hunk 2]
```diff
@@ -1,63 +1,87 @@
+// nolint
```
Comment (RajatPawar) [inline connector/params.go:1]: The `connector` namespace felt like a good place to have this, so moved the current auth/generic connectors out into `providers/generic`. This also gets around import cycle issues (this has minimum disruption)

**Review by Cobalt0s — COMMENTED** [review]
The approach makes sense. 

`connectors.NewConnector` should be in different package or other code parts should be moved, this is causing the cycling import. 

Does it mean that lets say `salesforce` package would move under `internal` or we could still initialize salesforce directly? In the second case parameters validation would be bypassed.

Let me know what actions do I need to take.

___________________________
[what] As a crazy idea we could design "special/small purpose" connector components that each would implement some unique interface, each associated with one of parameters (ex: WorkspaceHolderInterface, MultiModulesInterface).
[how] Then during constructor time internal utility would go over those interfaces asking if connector implements it, this way it will create a list of "required parameters" by relying on composition (they are implied because each parameter matches interface, thats why idea is crazy ahaha).
[why] This would eliminate `mustAuthenticatedClient, mustMetadata, etc`. Embedding such structs into Connector struct:
```
type Connector struct {
 component.Workspace      // inlcuding this would trigger constructor to call mustWorkspace
 component.MultiModules
}
```

**Review by eberle1080 — COMMENTED** [review]
Overall looks good. I'd approve but this is still in draft state.

**Comment by RajatPawar** [comment]
> The approach makes sense.
> 
> `connectors.NewConnector` should be in different package or other code parts should be moved, this is causing the cycling import.
> 
> Does it mean that lets say `salesforce` package would move under `internal` or we could still initialize salesforce directly? In the second case parameters validation would be bypassed.
> 
> Let me know what actions do I need to take.
> 
> [what] As a crazy idea we could design "special/small purpose" connector components that each would implement some unique interface, each associated with one of parameters (ex: WorkspaceHolderInterface, MultiModulesInterface). [how] Then during constructor time internal utility would go over those interfaces asking if connector implements it, this way it will create a list of "required parameters" by relying on composition (they are implied because each parameter matches interface, thats why idea is crazy ahaha). [why] This would eliminate `mustAuthenticatedClient, mustMetadata, etc`. Embedding such structs into Connector struct:
> 
> ```
> type Connector struct {
>  component.Workspace      // inlcuding this would trigger constructor to call mustWorkspace
>  component.MultiModules
> }
> ```

One could still initialize salesforce directly, so like you said, the validation is bypassed. The current cost we have to pay for validation is quite high - there's types we need to maintain & repeat across packages. By establishing `connector.NewConnector` as the entry point, we make an acceptable compromise for now

File: connector/params.go [hunk 3]
```diff
@@ -1,63 +1,87 @@
...
-)
+// Parameters can be used to pass input parameters to the connector. The json tags help with validation. Do not
+// remove these without updating the validation logic.
+type Parameters struct {
+	HTTPClient *http.Client
+	Module     common.ModuleID
 
-type Option = func(*parameters)
+	// AuthenticatedClient is a client for the connector that knows how to handle authentication for the provider.
+	AuthenticatedClient common.AuthenticatedHTTPClient
```
[the remark this answers is not in the document]
Comment (RajatPawar) [reply connector/params.go:21]: These are common parameters across all connectors, so some connectors might want to accept an `HTTPClient`. Most of them want `AuthenticatedClient`, but this adds optionality. I might have added this for the mock connector, but don't think anyone's really using it, so let me take it out until we really need it.

File: connector/new.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+		conn, err = buildConnector(newKlaviyoConnector, parameters, mustAuthenticatedClient)
+	case providers.CustomerJourneysApp:
+		conn, err = buildConnector(newCustomerJourneysAppConnector, parameters, mustAuthenticatedClient)
+	case providers.ConstantContact:
+		conn, err = buildConnector(newConstantContactConnector, parameters, mustAuthenticatedClient)
+	case providers.Keap:
+		conn, err = buildConnector(newKeapConnector, parameters, mustAuthenticatedClient)
+	default:
+		conn, err = newGenericConnector(provider, parameters, mustAuthenticatedClient)
+	}
```
[the remark this answers is not in the document]
Comment (RajatPawar) [reply connector/new.go:90]: Makes sense!

**Comment by RajatPawar** [comment]
@eberle1080 Marked it as ready for review, can you approve this?

**Review by eberle1080 — APPROVED** [review]

File: connector/new.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+	return generic.NewConnector(
+		provider,
+		generic.WithAuthenticatedClient(parameters.AuthenticatedClient),
+		generic.WithWorkspace(parameters.Workspace),
+	)
+}
+
+func newSalesforceConnector(parameters Parameters) (connectors.Connector, error) {
```
[the remark this answers is not in the document]
Comment (RajatPawar) [reply connector/new.go:131]: I am still using the old `package.WithParams()` styled options to initialize the connectors to make this change as small as possible. The change you described should be possible when I remove those in favor of the common `Parameters` in another PR.

File: connector/params.go [hunk 6]
```diff
@@ -1,63 +1,87 @@
...
+	}
 }
 
-// WithClient sets the http client to use for the connector.
-func WithClient(ctx context.Context, client *http.Client,
-	config *oauth2.Config, token *oauth2.Token, opts ...common.OAuthOption,
-) Option {
-	return func(params *parameters) {
-		params.WithOauthClient(ctx, client, config, token, opts...)
+func mustMetadata(params *Parameters) {
```
[the remark this answers is not in the document]
Comment (RajatPawar) [reply connector/params.go:58]: Agree, this is an intermediate state to make the transition as small as possible, will follow up

**Comment by jlimatampersand** [comment]
@RajatPawar Reverting this PR because this breaks server without dependency update work. Please do not merge PRs without resolving server code first.

**Pull request merged**